src/mongo_logger.ts

```diff
@@ -0,0 +1,214 @@
+  TRACE: 'trace',
+  OFF: 'off'
+} as const);
+
+/** @internal */
+export type SeverityLevel = typeof SeverityLevel[keyof typeof SeverityLevel];
+
+/** @internal */
+export const MongoLoggableComponent = Object.freeze({
+  COMMAND: 'command',
+  TOPOLOGY: 'topology',
+  SERVER_SELECTION: 'serverSelection',
+  CONNECTION: 'connection'
+} as const);
+
+/** @internal */
+export type MongoLoggableComponent =
+  typeof MongoLoggableComponent[keyof typeof MongoLoggableComponent];
+
+/** @internal */
+export interface MongoLoggerEnvOptions {
+  /** Severity level for command component */
+  MONGODB_LOG_COMMAND?: string;
+  /** Severity level for topology component */
+  MONGODB_LOG_TOPOLOGY?: string;
+  /** Severity level for server selection component */
+  MONGODB_LOG_SERVER_SELECTION?: string;
+  /** Severity level for CMAP */
+  MONGODB_LOG_CONNECTION?: string;
+  /** Default severity level to be if any of the above are unset */
+  MONGODB_LOG_ALL?: string;
+  /** Max length of embedded EJSON docs. Setting to 0 disables truncation. Defaults to 1000. */
+  MONGODB_LOG_MAX_DOCUMENT_LENGTH?: string;
+  /** Destination for log messages. Must be 'stderr', 'stdout'. Defaults to 'stderr'. */
+  MONGODB_LOG_PATH?: string;
+}
+
+/** @internal */
+export interface MongoLoggerMongoClientOptions {
+  /** Destination for log messages */
+  mongodbLogPath?: 'stdout' | 'stderr' | Writable;
+}
+
+/** @internal */
+export interface MongoLoggerOptions {
+  componentSeverities: {
+    /** Severity level for command component */
+    command: SeverityLevel;
+    /** Severity level for topology component */
+    topology: SeverityLevel;
+    /** Severity level for server selection component */
+    serverSelection: SeverityLevel;
+    /** Severity level for connection component */
+    connection: SeverityLevel;
+    /** Default severity level to be used if any of the above are unset */
+    default: SeverityLevel;
+  };
+
+  /** Max length of embedded EJSON docs. Setting to 0 disables truncation. Defaults to 1000. */
+  maxDocumentLength: number;
+  /** Destination for log messages. */
+  logDestination: Writable;
+}
+
+/**
+ * Parses a string as one of SeverityLevel
+ *
+ * @param s - the value to be parsed
+ * @returns one of SeverityLevel if value can be parsed as such, otherwise null
+ */
+function parseSeverityFromString(s?: string): SeverityLevel | null {
+  const validSeverities: string[] = Object.values(SeverityLevel);
+  const lowerSeverity = s?.toLowerCase();
+
+  if (lowerSeverity != null && validSeverities.includes(lowerSeverity)) {
+    return lowerSeverity as SeverityLevel;
+  }
+
+  return null;
+}
+
+/**
+ * Parses a string to be a number greater than or equal to 0 for maxDocumentLength.
+ *
+ * @param s - the value to be parsed
+ * @returns the int value parsed or 1000 if the value could not be parsed
+ */
+function parseMaxDocumentLength(s?: string): number {
+  try {
+    const maxDocumentLength = getUint('MONGODB_LOG_MAX_DOCUMENT_LENGTH', s);
```

Comment:
Since we're not actually using the validation that this function provides (the catch squashes the thrown error) could the helper be put back into connection_string.ts and we use Number.parseInt + Number.isNaN here?

Reply:
do we know for a fact that we won't validate this in the future? (particularly, when it comes from client options)

Reply:
Well, we sort of use the validation.  `getUInt` also asserts that it's non-negative.  We currently catch+default instead of throwing but I think that it still makes sense to rely on the function we already have that parses uints from strings

Reply:
We might validate the options when they come from the client options but that will happen in the options parser. It's just about using try/catch for something that's a condition check, it's not the control flow you'd expect when you need to make sure something is greater than zero